beefy-gadget/src/worker.rs

```diff
@@ -402,3 +384,133 @@ where
+	let candidate = best_beefy + min_delta.max(diff.next_power_of_two()).into();
+
+	trace!(
+		target: "beefy",
+		"🥩 should_vote_on: #{:?}, diff: {:?}, next_power_of_two: {:?}, next_block_to_vote_on: #{:?}",
+		number,
+		diff,
+		diff.next_power_of_two(),
+		candidate,
+	);
+
+	candidate
+}
+
+#[cfg(test)]
+mod tests {
+	use super::vote_candidate;
+	use sp_runtime::testing::{Block, ExtrinsicWrapper, Header};
+
+	type MockBlock = Block<ExtrinsicWrapper<u64>>;
+
+	macro_rules! block {
+		($n:expr) => {
+			Header::new_from_number($n).number;
+		};
+	}
+
+	#[test]
+	fn vote_on_min_block_delta() {
+		let c = vote_candidate::<MockBlock>(block!(1), block!(1), block!(0), 4);
+		assert_eq!(4, c);
+		let c = vote_candidate::<MockBlock>(block!(2), block!(2), block!(0), 4);
+		assert_eq!(4, c);
+		let c = vote_candidate::<MockBlock>(block!(3), block!(3), block!(0), 4);
+		assert_eq!(4, c);
+		let c = vote_candidate::<MockBlock>(block!(4), block!(4), block!(0), 4);
+		assert_eq!(4, c);
+
+		let c = vote_candidate::<MockBlock>(block!(4), block!(4), block!(4), 4);
+		assert_eq!(8, c);
+
+		let c = vote_candidate::<MockBlock>(block!(10), block!(10), block!(10), 4);
+		assert_eq!(14, c);
+		let c = vote_candidate::<MockBlock>(block!(11), block!(11), block!(10), 4);
+		assert_eq!(14, c);
+		let c = vote_candidate::<MockBlock>(block!(12), block!(12), block!(10), 4);
+		assert_eq!(14, c);
+		let c = vote_candidate::<MockBlock>(block!(13), block!(13), block!(10), 4);
+		assert_eq!(14, c);
+
+		let c = vote_candidate::<MockBlock>(block!(10), block!(10), block!(10), 8);
+		assert_eq!(18, c);
+		let c = vote_candidate::<MockBlock>(block!(11), block!(11), block!(10), 8);
+		assert_eq!(18, c);
+		let c = vote_candidate::<MockBlock>(block!(12), block!(12), block!(10), 8);
+		assert_eq!(18, c);
+		let c = vote_candidate::<MockBlock>(block!(13), block!(13), block!(10), 8);
+		assert_eq!(18, c);
+	}
+
+	#[test]
+	fn vote_on_power_of_two() {
+		let c = vote_candidate::<MockBlock>(block!(1008), block!(1008), block!(1000), 4);
+		assert_eq!(1008, c);
+
+		let c = vote_candidate::<MockBlock>(block!(1016), block!(1016), block!(1000), 4);
+		assert_eq!(1016, c);
+
+		let c = vote_candidate::<MockBlock>(block!(1032), block!(1032), block!(1000), 4);
+		assert_eq!(1032, c);
+
+		let c = vote_candidate::<MockBlock>(block!(1064), block!(1064), block!(1000), 4);
+		assert_eq!(1064, c);
+
+		let c = vote_candidate::<MockBlock>(block!(1128), block!(1128), block!(1000), 4);
+		assert_eq!(1128, c);
+
+		let c = vote_candidate::<MockBlock>(block!(1256), block!(1256), block!(1000), 4);
+		assert_eq!(1256, c);
+
+		let c = vote_candidate::<MockBlock>(block!(1512), block!(1512), block!(1000), 4);
+		assert_eq!(1512, c);
+
+		let c = vote_candidate::<MockBlock>(block!(1024), block!(1024), block!(0), 4);
+		assert_eq!(1024, c);
+	}
+
+	#[test]
+	fn vote_on_target_block() {
+		let c = vote_candidate::<MockBlock>(block!(1008), block!(1008), block!(1002), 4);
+		assert_eq!(1010, c);
```

Comment:
`candidate` blocks should not be greater than `best_grandpa` blocks. We must not vote for non-existent or non-finalized blocks.

Reply:
We will not. The expr `number == candidate` in the `should_vote_on()` does prevent that. Also running a local test node shows that we vote exactly as we should.

Reply:
The block number in the `assert_eq` is the `best_grandpa` number we would vote for.

Reply:
Yes, you're right apologies. I had incorrect understanding how the code works (i.e. what does it mean to "select a candidate") and this is why I was suggesting the changes. More context is available in this issue: https://github.com/paritytech/grandpa-bridge-gadget/issues/162